convex/http.ts

```diff
@@ -0,0 +1,108 @@
+import { httpRouter } from "convex/server";
+import { httpAction } from "./_generated/server";
+import OpenAI from "openai";
+import { internal } from "./_generated/api";
+import { Doc, Id } from "./_generated/dataModel";
+
+const http = httpRouter();
+
+const UPDATE_BATCH_SIZE = 20;
+
+http.route({
+  path: "/chat",
+  method: "POST",
+  handler: httpAction(async (ctx, request) => {
+    const body = await request.json();
+    const messageId: Id<"messages"> = body.messageId;
+    const messages: Doc<"messages">[] = body.messages;
+    const openai = new OpenAI();
+
+    // Create a TransformStream to handle streaming data
+    let { readable, writable } = new TransformStream();
+    let writer = writable.getWriter();
+    const textEncoder = new TextEncoder();
+
+    const streamData = async () => {
+      let content = "";
+      const stream = await openai.chat.completions.create({
+        model: "gpt-3.5-turbo",
+        messages: [
+          {
+            role: "system",
+            content:
+              "You are a really excited bot in a group chat responding to q's.",
+          },
+          ...messages.map(({ body, author }) => ({
+            role:
+              author === "ChatGPT" ? ("assistant" as const) : ("user" as const),
+            content: body,
+          })),
+        ],
+        stream: true,
+      });
+
+      // loop over the data as it is streamed and write to the writeable
+      // also periodically update the message in the database
+      let partsProcessed = 0;
+      for await (const part of stream) {
+        const text = part.choices[0]?.delta?.content || "";
+        content += text;
+        partsProcessed += 1;
+        await writer.write(textEncoder.encode(text));
+        if (partsProcessed % UPDATE_BATCH_SIZE === 0) {
```

Comment:
in case you want it, for llama chat I used this to decide when to send an update:
```ts
export function hasDelimeter(response: string) {
  return (
    response.includes("\n") ||
    response.includes(".") ||
    response.includes("?") ||
    response.includes("!") ||
    response.includes(",") ||
    response.length > 100
  );
}
```